gitlab_runner_config.py

```diff
@@ -200,113 +263,35 @@ def update_runner_config(config_template, config_file, internal_config):
-                    runner_config[runner_type] = runner
-                except KeyError:
-                    # we may have picked up a runner which doesn't match our
-                    # host, skip it
-                    pass
-    else:
-        try:
-            # ensure tokens are still valid, otherwise, delete the runner and
-            # register it again
-            data_file = os.path.join(prefix, "runner-data.json")
-            with open(data_file, "r") as fh:
-                changed = False
-                runner_config = json.load(fh)
-                for runner_type, data in runner_config.items():
-                    data = data or {}
-                    token = data.get("token", "")
-                    if not token or not valid_runner_token(api_url, token):
-                        # no refresh endpoint...delete and re-register
-                        if token:
-                            delete_runner(api_url, token)
-                        runner_config[runner_type] = register_runner(
-                            api_url, admin_token, runner_type, runner_type=runner_type
-                        )
-                        changed = True
-                if changed:
-                    with open(data_file, "w") as fh:
-                        fh.write(json.dumps(runner_config, sort_keys=True, indent=4))
-        except FileNotFoundError:
-            # defaults to creating both a shell and batch runner
-            runner_config = {
-                t: register_runner(
-                    api_url, admin_token, t, generate_tags(runner_type=t)
-                )
-                for t in ["shell", "batch"]
-            }
-            with open(data_file, "w") as fh:
-                fh.write(json.dumps(runner_config, sort_keys=True, indent=4))
+    runners = {}
 
-    update_runner_config(config_template, config_file, runner_config)
+    for filename in os.listdir(templates):
+        if filename.endswith(".toml"):
+            runners.update(read_runner(templates, filename, clients))
 
+    config_file = os.path.join(prefix, "config.toml")
+    config_template = os.path.join(prefix, "config.template")
+    update_runner_config(config_template, config_file, runners)
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(description="On the fly runner config")
     parser.add_argument(
         "-p",
         "--prefix",
         default="/etc/gitlab-runner",
-        help="""The runner config directory prefix""",
-    )
-    parser.add_argument(
-        "--api-url", default="http://localhost:8080/api/v4", help="""Gitlab API URL"""
+        help="The runner config directory prefix"
     )
     parser.add_argument(
-        "--stateless",
-        action="store_true",
-        help="""If used, disables writing runner-data.json and must query
-        Gitlab directly for state.
-        """,
+        "-t",
+        "--templates",
+        default="/etc/gitlab-runner/main",
```

Comment:
Rather than specify this as different path altogether, let's assume that whatever template directory is being used here exists under `prefix`. I would also suggest renaming to `instance` or something like that.

This will allow for a simpler invocation of `systemd`. The execution line would change to:

`ExecStartPre=gitlab_runner_auth.py --prefix /etc/gitlab-runner --instance %i`

And the instance would come from invoking `systemctl` to start another process:

`systemctl start gitlab-runner@bar.service # bar would populate %i above`.

In the absence of invoking with `@`, the default behavior would be to look for a folder `main` so the default arg route you have above would work for that (sans absolute path).

See this post for more details: https://stackoverflow.com/a/48232848